src/layers/loss.h

```diff
@@ -437,14 +437,46 @@ class SequenceUnlikelihoodLoss : public CrossEntropyLoss {
 
 /**
  * @brief Cross entropy in rescorer used for computing sentences-level log probabilities
+ *
+ * This class differs from CrossEntropy in the different 'axes' setting, and that label smoothing
+ * is disabled.
  */
 class RescorerLoss : public CrossEntropyLoss {
 public:
   // sentence-wise CE, hence reduce only over time axis
-  // This class differs from CrossEntropy in the different 'axes' setting, and that label smoothing is disabled.
   RescorerLoss() : CrossEntropyLoss(/*axes=*/{-3} /*time axis*/, /*smoothing=*/0.f, /*factorWeight=*/1.0f) {}
 };
 
+/**
+ * @brief Cross entropy in rescorer used for computing word-level log probabilities
+ *
+ * This class differs from CrossEntropy that it does not reduce the loss over any axis (counts are
+ * still reduced to get sentence lengths), and that label smoothing is disabled.
+ */
+class RescorerUnreducedLoss : public CrossEntropyLoss {
+public:
+  // word-level CE, hence do not reduce over any axis
+  RescorerUnreducedLoss() : CrossEntropyLoss(/*axes=*/{}, /*smoothing=*/0.f, /*factorWeight=*/1.0f) {}
+
+  virtual RationalLoss apply(Logits logits, const Words& labels,
+                             Expr mask = nullptr, Expr labelWeights = nullptr) override {
+    auto loss = CrossEntropyLoss::compute(logits, labels, mask, labelWeights);
+    ABORT_IF(!mask, "Word-level CE from rescorer must have mask");
+    return reduceLabels(loss, mask);  // reduce labels to get sentence lengths
+  }
+
+  RationalLoss reduceLabels(Expr loss, Expr labels) {
```

Comment:
This is just a utlity function, right? Not meant for inheriting etc. So I guess that should either be private or actually just merged with `apply()`. I don't see a good reason for having that split into two functions?

Reply:
And since we are already talking about merging. Would it make sense to just merge the two classes `RescorerLoss` and `RescorerUnreducedLoss` and just pass a boolean argument `bool wordLevelScores` in order to tell what is being returned and what behavior is expected?

Reply:
You are totally right, this will be much cleaner. Thanks!

Reply:
Done